### PR TITLE
Fix RTC implementation for the Data Register (DR)

### PR DESCRIPTION
**Pull request by andreeaflorescu**
The DR register needs to return the current time + any adjustments made by the driver. With the previous implementation, the time was in the 1970.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
To test this with the vmm-reference ARM poc (on an aarch64 machine):

```bash
# Fetch the vmm-reference arm poc code
git clone https://github.com/andreeaflorescu/vmm-reference.git
cd vmm-reference
git checkout arm_poc

# Download a PE image
wget https://vmm-reference-test-resources.s3.eu-central-1.amazonaws.com/v2/kernel/pe-hello-busybox

 # Run the vmm-reference and check that the time is correct
 cargo run -- --kernel path=pe-hello-busybox --memory size_mib=10000
```

File: src/rtc_pl031.rs [hunk 1]
```diff
@@ -322,6 +334,21 @@ mod tests {
...
+        // The problem is that the time in the guest would show up as in the 1970s.
+        let mut rtc = RTC::new();
+        let expected_time = get_epoch_time();
+
+        let mut actual_time = [0u8; 4];
+        rtc.read(RTCDR, &mut actual_time);
+        // Check that the difference between the current time, and the time read from the
+        // RTC device is never bigger than one second. This should hold true irrespective of
+        // scheduling.
+        assert!(u32::from_le_bytes(actual_time) - expected_time <= 1);
```
Comment (lauralt) [inline src/rtc_pl031.rs:349]: If these are integers, should we just check for equality (as it is already done in the other tests)?

File: src/rtc_pl031.rs [hunk 2]
```diff
@@ -135,6 +135,18 @@ pub struct RTC<EV: RTCEvents> {
...
+fn get_epoch_time() -> u32 {
+    let epoch_time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        // This expect should never fail because UNIX_EPOCH is in 1970,
+        // and the only possible failure is if `now` time is before UNIX EPOCH.
+        .expect("SystemTime::duration_since failed");
+    // The following conversation is save because u32::MAX would correspond to
+    // year 2106, and the RTC works exclusively with 32-bit registers, so we
+    // wouldn't be able to fit more than this in the LR (Load Register).
+    epoch_time.as_secs() as u32
```
Comment (lauralt) [inline src/rtc_pl031.rs:147]: Hmm, it's still not clear to me why this conversion is safe. If we'd be in 2107, wouldn't as_secs() be > u32::MAX? Should we simplify this comment saying we don't expect to use this after 2106 or am I missing something?

File: src/rtc_pl031.rs [hunk 3]
```diff
@@ -166,8 +178,8 @@ impl<EV: RTCEvents> RTC<EV> {
             // Counts up from 1 on reset at 1Hz (emulated).
             counter: Instant::now(),
 
-            // The load register is initialized to zero.
-            lr: 0,
+            // The load register is initialized to UNIX epoch time.
+            lr: get_epoch_time(),
```
[the remark this answers is not in the document]
Comment (lauralt) [reply src/rtc_pl031.rs:182]: This comment applies to `test_control_register` as well.

**Comment by dodsonmg** [comment]
I may have misunderstood the spec, but it explicitly has a reset value of `0x00000000` for the data and load registers, which is why I wrote the RTC such that on reset it count up from 0.

I had assumed the bus or the application would set the initial value as desired (i.e., current time relative to `UNIX_EPOCH`, if using the RTC to provide wall-clock time).

File: src/rtc_pl031.rs (hunk 2 above)
Comment (andreeaflorescu) [reply src/rtc_pl031.rs:147]: Epoch time would only not fit in the u32 if we were in year 2106. Since that is a reaaally long time from now, this conversion is safe. We would also not be able to use RTC at that point (at least this one), because the register that holds this value is an u32.

File: src/rtc_pl031.rs (hunk 1 above)
Comment (andreeaflorescu) [reply src/rtc_pl031.rs:349]: If my understanding is correct, we cannot check for equality here because there might be 1 second elapsed between RTC::new() and the call on the next line `get_epoch_time`. If this happens, then the test will intermittently fail, and thus is should be safe to assume that that at most 1 second can pass between the two calls.

**Comment by andreeaflorescu** [comment]
> I may have misunderstood the spec, but it explicitly has a [reset value](https://developer.arm.com/documentation/ddi0224/c/Programmers-model/Summary-of-RTC-registers?lang=en) of `0x00000000` for the data and load registers, which is why I wrote the RTC such that on reset it count up from 0.
> 
> I had assumed the bus or the application would set the initial value as desired (i.e., current time relative to `UNIX_EPOCH`, if using the RTC to provide wall-clock time).

@dodsonmg the reset value is indeed 0 (and I think I left that unchanged). My intention was to change the initial value of the load register to unix epoch because this is what is expected in Firecracker, and checking the code for the other VMMs, this is the expected value when the VMM boots:
- in Firecracker, by initializing the `tick_offset` to [the current time](https://github.com/firecracker-microvm/firecracker/blob/main/src/utils/src/time.rs#L143)
- in Crosvm (where they do not have the reset functionality implemented), where they always return the epoch time in https://chromium.googlesource.com/chromiumos/platform/crosvm/+/refs/heads/main/devices/src/pl030.rs#131

The implementation that we have here has more functionality looking at the others, as it is also (correctly) implementing the reset.

File: src/rtc_pl031.rs (hunk 2 above)
Comment (lauralt) [reply src/rtc_pl031.rs:147]: Can we rephrase it to something like:

```
    // The following conversion is safe because u32::MAX would correspond to
    // year 2106, and we don't expect to use it afterwards. We are doing this
    // conversion because the RTC works exclusively with 32-bit registers, so we
    // wouldn't be able to fit more than this in the LR (Load Register).
```
?
The way I was reading the comment, the fact that the RTC works with 32bit registers is a reason why the conversion is safe, and if I get things right, this is not true.

**Comment by andreeaflorescu** [comment]
Looking a bit more at the documentation, it seems like the register that is not correctly implemented is the Data Register (as the issue stated initially), and not the Load Register (which is what I was saying in this PR).

As @dodsonmg also said, the LR should start counting from 0.

To fix this problem, there are quite a few changes needed, so I am converting this back to an RFC.

**Comment by dodsonmg** [comment]
> Looking a bit more at the documentation, it seems like the register that is not correctly implemented is the Data Register

I'm still confused by this.

My understanding of the spec is that if you read the data register an instant after power on, you would get a value of 0.  The purpose of the RTC is to give you a 1Hz signal and a limited alarm functionality.  If you want the RTC to approximate wall time, then the first thing that should be done after instantiating the RTC is to load the current offset from `UNIX_EPOCH` via the load register.

> To fix this problem, there are quite a few changes needed, so I am converting this back to an RFC.

If, however, the users of this emulated RTC are expecting different behaviour (per your references to firecracker and crosvm above), then I'm not sure it matters how it gets implemented, and the proposal here seems as good as any other.  Put another way, if we're going to deviate from the spec behaviour (as I understand it... which might be wrong!) then the simplest (or easiest to document) implementation seems to be the way forward.

File: src/rtc_pl031.rs [hunk 4]
```diff
@@ -8,7 +8,7 @@
 //! counting for a programmed number of cycles of a real-time clock input.
 //!
 use std::sync::Arc;
-use std::time::Instant;
+use std::time::{Instant, SystemTime, UNIX_EPOCH};
 
 // As you can see in
 //  https://static.docs.arm.com/ddi0224/c/real_time_clock_pl031_r1p3_technical_reference_manual_DDI0224C.pdf
```
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/rtc_pl031.rs]: Updated.

File: src/rtc_pl031.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/rtc_pl031.rs]: I would leave the example as is because the driver can still set the LR value (and UNIX Epoch is  a value that makes sense I think), and the example correctly checks that a value set in LR, can then be read from the register.

[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/rtc_pl031.rs]: I've updated `test_control_register`.

File: src/rtc_pl031.rs (hunk 2 above)
Comment (andreeaflorescu) [reply src/rtc_pl031.rs]: I've updated it. Let me know if it makes more sense now.

**Comment by andreeaflorescu** [comment]
@dodsonmg @lauralt I did quite a few updates to this implementation.
In summary:
- DR returns the current time + the offset (which is computed as defined in the spec)
- since we do not implement the alarm functionality, and we implement DR as current time + offset, the counter is no longer useful, and I removed it. If we decide to implement the alarm at some point, we'll need to see how assert the interrupt

Can you p.lease take another look?

File: src/rtc_pl031.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply src/rtc_pl031.rs]: Looks like with the new changes my comment doesn't make sense anymore. [This one](https://github.com/rust-vmm/vm-superio/compare/c0491934e3ff2ff3e3fa3efc22b4254fb23ed5ee..f66c624522ba0acdb5a7e5bddfec4ba7f8205c84#diff-20e2d11876ae9564e0e163379ffa125952680cfc6b71daf9644f52a892a23b58R178) should be updated though.

[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/rtc_pl031.rs]: Done.

**Review by lauralt — APPROVED** [review]

**Review by alexggh — APPROVED** [review]

**Review by aghecenco — APPROVED** [review]
Rubber stamp :heavy_check_mark:
